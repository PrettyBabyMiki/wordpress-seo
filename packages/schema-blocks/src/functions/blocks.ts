import { dispatch, select } from "@wordpress/data";

/**
 * Gets attributes of inner blocks.
 *
 * @param clientId The client ID of the parent block.
 * @param blocks   A mapping of block name to attribute key.
 *
 * @returns An array contain block names and values.
 */
export function getInnerBlocksAttributes( clientId: string, blocks: Record<string, string> ): { name: string; value: unknown }[] {
	let innerBlocks = select( "core/block-editor" ).getBlock( clientId ).innerBlocks;
	innerBlocks     = innerBlocks.filter( block => block.name in blocks );

	const values = [];

	for ( const block of innerBlocks ) {
		const key   = blocks[ block.name ];
		values.push( { name: block.name, value: block.attributes[ key ] } );
	}

	return values;
}

/**
 * Removes a block from the editor.
 *
 * @param {string} clientId The client id of the block to remove.
 */
export function removeBlock( clientId: string ) {
	dispatch( "core/block-editor" ).removeBlock( clientId );
}
