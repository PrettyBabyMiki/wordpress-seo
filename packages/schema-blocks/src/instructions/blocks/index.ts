import "./Block";
import "./ClassName";
import "./InnerBlocks";
import "./RichText";
import "./SidebarDuration";
import "./SidebarInput";
import "./SidebarSelect";
import "./VariableTagRichText";
