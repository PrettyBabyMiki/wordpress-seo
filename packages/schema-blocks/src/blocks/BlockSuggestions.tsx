import { ReactElement } from "react";
import { BlockInstance, createBlock } from "@wordpress/blocks";
import { createElement } from "@wordpress/element";
import { getInnerblocksByName, insertBlock } from "../functions/innerBlocksHelper";
import { getBlockType } from "../functions/BlockHelper";
import { PanelBody } from "@wordpress/components";
import { SuggestedBlockProperties } from "../core/validation/SuggestedBlockProperties";

type BlockSuggestionAddedDto = {
	blockTitle: string;
}

type BlockSuggestionDto = {
	blockTitle: string;
	blockName: string;
	blockClientId: string;
}

/**
 * Renders a block suggestion with the possibility to add one.
 *
 * @param {string}   blockTitle    The title to show.
 * @param {string}   blockName     The name of the block to add.
 * @param {string}   blockClientId The clientId of the target to add the block to.
 *
 * @returns {ReactElement} The rendered block suggestion.
 */
function BlockSuggestion( { blockTitle, blockName, blockClientId }: BlockSuggestionDto ): ReactElement {
	/**
	 * Onclick handler for the remove block.
	 */
	const addBlockClick = () => {
		const blockToAdd = createBlock( blockName );
		insertBlock( blockToAdd, blockClientId );
	};

	return (
		<li className="yoast-block-suggestion">
			{ blockTitle }
			<button className="yoast-block-suggestion-button" onClick={ addBlockClick }> Add </button>
		</li>
	);
}

/**
 * Renders a block suggestion that is already added
 *
 * @param {string} blockTitle The block title.
 *
 * @returns {ReactElement} The rendered element.
 */
function BlockSuggestionAdded( { blockTitle }: BlockSuggestionAddedDto ): ReactElement {
	return (
		<li className="yoast-block-suggestion yoast-block-suggestion--added">
			{ blockTitle }
			<span className="yoast-block-suggestion-checkmark"> OK </span>
		</li>
	);
}

/**
 * Renders a list with the required/recommended block names and an button to add/remove one.
 *
 * @param {string} sidebarTitle                The title of the sidebar section.
 * @param {BlockInstance} block                The block to render the list for.
 * @param {SuggestedBlockProperties[]} blocks The required/recommended blocks.
 *
 * @returns {ReactElement} The rendered block.
 */
export default function RequiredBlocks( sidebarTitle: string, block: BlockInstance, blocks: SuggestedBlockProperties[] ): ReactElement {
	// Retrieve a list with names.
	const relevantBlockNames = blocks
		.filter( relevantBlock => typeof getBlockType( relevantBlock.name ) !== "undefined" )
		.map( relevantBlock => relevantBlock.name );

	// When there are no required blocknames, just return.
	if ( relevantBlockNames.length === 0 ) {
		return null;
	}

	const findPresentBlocks = getInnerblocksByName( block, relevantBlockNames );
	const presentBlockNames = findPresentBlocks.map( presentBlock => presentBlock.name );

	return (
		<PanelBody key={ block.clientId }>
			<div className="yoast-block-sidebar-title">{ sidebarTitle }</div>
			<ul className="yoast-block-suggestions">
				{
					relevantBlockNames.map( ( blockName: string, index: number ) => {
						const blockType = getBlockType( blockName );

						if ( presentBlockNames.includes( blockName ) ) {
							return <BlockSuggestionAdded key={ index } blockTitle={ blockType.title } />;
						}

						return <BlockSuggestion
							key={ index }
							blockTitle={ blockType.title }
							blockName={ blockName }
							blockClientId={ block.clientId }
						/>;
					} )
				}
			</ul>
		</PanelBody>
	);
}
