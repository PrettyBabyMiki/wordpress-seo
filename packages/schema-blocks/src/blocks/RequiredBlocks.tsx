import { ReactElement } from "react";
import { BlockInstance, createBlock } from "@wordpress/blocks";
import { createElement } from "@wordpress/element";
import { RequiredBlock } from "../instructions/blocks/dto";

import { getInnerblocksByName, insertBlockToInnerBlock } from "../functions/innerBlocksHelper";
import { getBlockType } from "../functions/blocks";
import { PanelBody } from "@wordpress/components";

/**
 * Renders a list with the required block names and an button to add/remove one.
 *
 * @param {BlockInstance}   block          The block to render the list for.
 * @param {RequiredBlock[]} requiredBlocks The required blocks.
 *
 * @returns {ReactElement} The rendered block.
 */
export default function RequiredBlocks( block: BlockInstance, requiredBlocks: RequiredBlock[] ): ReactElement {
	// Retrieve a list with names.
	const requiredBlockNames = requiredBlocks.map( ( requiredBlock ) => {
		return requiredBlock.name;
	} );

	const findPresentBlocks = getInnerblocksByName( block, requiredBlockNames );
	const presentBlockNames = findPresentBlocks.map( ( presentBlock ) => {
		return presentBlock.name;
	} );

	const requiredBlockItems: ReactElement[] = [];

	requiredBlockNames.forEach( ( requiredBlockName: string ) => {
		const blockType = getBlockType( requiredBlockName );

		if ( typeof blockType === "undefined" ) {
			return;
		}

		if ( presentBlockNames.includes( requiredBlockName ) ) {
			requiredBlockItems.push(
				(
					<li className="yoast-block-suggestion yoast-block-suggestion--added">
						{ blockType.title }
						<span className="yoast-block-suggestion-checkmark">x</span>
					</li>
				),
			);

			return;
		}

		/**
		 * Onclick handler for the remove block.
		 */
		const addBlockClick = () => {
			const blockToAdd = createBlock( requiredBlockName );
			insertBlockToInnerBlock( blockToAdd, block.clientId );
		};

		requiredBlockItems.push(
			(
				<li className="yoast-block-suggestion">
					{ blockType.title }
					<button className="yoast-block-suggestion-button" onClick={ addBlockClick }>Add</button>
				</li>
			),
		);
	} );

	if ( requiredBlockItems.length === 0 ) {
		return null;
	}

	return (
		<PanelBody>
			<div className="yoast-block-sidebar-title">Required blocks</div>
			<ul className="yoast-block-suggestions">
				{ requiredBlockItems }
			</ul>
		</PanelBody>
	);
}
