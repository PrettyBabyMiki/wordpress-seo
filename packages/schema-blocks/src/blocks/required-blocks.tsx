import { ReactElement } from "react";
import { BlockInstance, createBlock } from "@wordpress/blocks";
import { createElement } from "@wordpress/element";
import { RequiredBlock } from "../instructions/blocks/dto";

import { getInnerblocksByName, insertBlockToInnerBlock } from "../functions/innerBlocksHelper";
import { getBlockType, removeBlock } from "../functions/blocks";
import { PanelBody, PanelRow } from "@wordpress/components";

/**
 * Renders a required block button.
 *
 * @param {BlockInstance} block             The block to render the button for.
 * @param {string}        requiredBlockName The required block name.
 * @param {boolean}       blockDoesExists   Is the block already added.
 *
 * @returns {ReactElement} The rendered button.
 */
const RequiredBlockButton = ( block: BlockInstance, requiredBlockName: string, blockDoesExists: boolean ): ReactElement => {
	if ( blockDoesExists ) {
		/**
		 * Onclick handler for the remove block.
		 */
		const removeBlockClick = () => {
			const blocksToRemove = getInnerblocksByName( block, [ requiredBlockName ] );
			blocksToRemove.forEach( ( blockToRemove ) => {
				removeBlock( blockToRemove.clientId );
			} );
		};

		return <button className="yoast-block-suggestion-button" onClick={ removeBlockClick }>Remove</button>;
	}

	/**
	 * Onclick handler for the remove block.
	 */
	const addBlockClick = () => {
		const blockToAdd = createBlock( requiredBlockName );
		insertBlockToInnerBlock( blockToAdd, block.clientId );
	};

	return <button className="yoast-block-suggestion-button" onClick={ addBlockClick }>Add</button>;
};

/**
 * Renders a list with the required block names and an button to add/remove one.
 *
 * @param {BlockInstance}   block          The block to render the list for.
 * @param {RequiredBlock[]} requiredBlocks The required blocks.
 *
 * @returns {ReactElement}
 */
export default function RequiredBlocks( block: BlockInstance, requiredBlocks: RequiredBlock[] ): ReactElement {
	// Retrieve a list with names.
	const requiredBlockNames = requiredBlocks.map( ( requiredBlock ) => {
		return requiredBlock.name;
	} );

	const findPresentBlocks = getInnerblocksByName( block, requiredBlockNames );
	const presentBlockNames = findPresentBlocks.map( ( presentBlock ) => {
		return presentBlock.name;
	} );

	const requiredBlockItems: ReactElement[] = [];

	requiredBlockNames.forEach( ( requiredBlockName: string ) => {
		const blockType = getBlockType( requiredBlockName );

		if ( typeof blockType === "undefined" ) {
			return;
		}

		const button = RequiredBlockButton( block, requiredBlockName, presentBlockNames.includes( requiredBlockName ) );

		requiredBlockItems.push( <li className="yoast-block-suggestion">{ blockType.title }{ button }</li> );
	} );

	return (
		<PanelBody title="Required blocks">
			<PanelRow>
				<ul className="yoast-block-suggestions">
					{ requiredBlockItems }
				</ul>
			</PanelRow>
		</PanelBody>
	);
}
