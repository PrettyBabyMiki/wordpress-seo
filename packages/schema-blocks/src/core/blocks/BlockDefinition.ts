import { createElement, Fragment } from "@wordpress/element";
import { registerBlockType, BlockConfiguration, BlockEditProps, BlockSaveProps } from "@wordpress/blocks";
import { InspectorControls } from "@wordpress/block-editor";

import BlockInstruction from "./BlockInstruction";
import Definition from "../Definition";
import BlockRootLeaf from "../../leaves/blocks/BlockRootLeaf";
import parse from "../../functions/blocks/parse";

export interface RenderEditProps extends BlockEditProps<Record<string, unknown>> {
	clientId?: string;
}

export interface RenderSaveProps extends BlockSaveProps<Record<string, unknown>> {
	clientId?: string;
}

export type MutableBlockConfiguration = {
	-readonly [K in keyof BlockConfiguration]: BlockConfiguration[K]
}

/**
 * BlockDefinition class
 */
export default class BlockDefinition extends Definition {
	public static separatorCharacters = [ "@", "#", "$", "%", "^", "&", "*", "(", ")", "{", "}", "[", "]" ];
	public static parser = parse;

	public instructions: Record<string, BlockInstruction>;
	public tree: BlockRootLeaf;

	/**
	 * Renders editing the block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	edit( props: RenderEditProps ): JSX.Element {
		// Take the children directly to avoid creating too many Fragments.
		const elements = this.tree.children.map( ( leaf, i ) => leaf.edit( props, i ) ).filter( e => e !== null );

		const sidebarElements = Object.values( this.instructions )
			.map( ( instruction, i ) => instruction.sidebar( props, i ) )
			.filter( e => e !== null );
		if ( sidebarElements.length > 0 ) {
			const sidebar = createElement( InspectorControls, null, sidebarElements );
			elements.unshift( sidebar );
		}

		if ( elements.length === 1 ) {
			return elements[ 0 ] as JSX.Element;
		}

		return createElement( Fragment, null, elements );
	}

	/**
	 * Renders saving the block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	save( props: RenderSaveProps ): JSX.Element {
		return this.tree.save( props );
	}

	/**
	 * Registers the block with Gutenberg.
	 */
	register(): void {
		const configuration = this.configuration() as MutableBlockConfiguration;

		const name = configuration.name as string;
		delete configuration.name;

		configuration.edit = props => this.edit( props );
		configuration.save = props => this.save( props );

		registerBlockType( name, configuration );
	}

	/**
	 * Checks if the definition block is valid.
	 *
	 * @returns {boolean} True if the definition block is valid, False if the block contains errors.
	 */
	valid(): boolean {
		return true;
	}
}
