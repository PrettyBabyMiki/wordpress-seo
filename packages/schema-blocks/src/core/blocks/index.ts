import BlockDefinition from "./BlockDefinition";
import BlockInstruction from "./BlockInstruction";
import { BlockInstructionClass } from "./BlockInstructionClass";
import BlockLeaf from "./BlockLeaf";
import * as BlockDefinitionRepository from "./BlockDefinitionRepository";
import { ValidatingBlockInstruction } from "./ValidatingBlockInstruction";

export {
	BlockDefinition,
	BlockInstruction,
	BlockInstructionClass,
	BlockLeaf,
	BlockDefinitionRepository,
	ValidatingBlockInstruction,
};
