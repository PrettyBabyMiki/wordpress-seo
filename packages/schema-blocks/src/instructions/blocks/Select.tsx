import BlockInstruction from "../../core/blocks/BlockInstruction";
import { RenderEditProps, RenderSaveProps } from "../../core/blocks/BlockDefinition";
import { createElement, ReactElement } from "@wordpress/element";
import { BlockConfiguration, BlockInstance } from "@wordpress/blocks";
import { attributeExists, attributeNotEmpty } from "../../functions/validators";
import { arrayOrObjectToOptions } from "../../functions/select";

/**
 * Select (a drop-down box) instruction.
 */
export default class Select extends BlockInstruction {
	public options: {
		/**
		 * The attribute name the value selected in the select control should be saved as.
		 */
		name: string;
		/**
		 * The available options within the select control.
		 */
		options: string[] | Record<string, string>;
		/**
		 * The label that should be shown before this select control.
		 */
		label: string;
		/**
		 * If it is required that a value is selected.
		 */
		required?: boolean;
	};

	/**
	 * Renders saving the element.
	 *
	 * @param props The props.
	 *
	 * @returns {JSX.Element} The element to render.
	 */
	save( props: RenderSaveProps ): ReactElement | string {
		return <p data-id={ this.options.name } data-value={ this.value( props ) }>
			<b>{ this.options.label }:</b> { this.label( props ) }
		</p>;
	}

	/**
	 * Returns the value of the selected option.
	 *
	 * @param props The render props.
	 *
	 * @returns The value of the selected option.
	 */
	protected value( props: RenderSaveProps | RenderEditProps ): string {
		return props.attributes[ this.options.name ] as string;
	}

	/**
	 * Returns the label of the selected option.
	 *
	 * @param props The render props.
	 *
	 * @returns The label of the selected option.
	 */
	protected label( props: RenderSaveProps | RenderEditProps ): string {
		if ( Array.isArray( this.options.options ) ) {
			// Value is the same as the label, so just return the value.
			return this.value( props );
		}

		for ( const [ label, value ] of Object.entries( this.options.options ) ) {
			if ( this.value( props ) === value ) {
				return label;
			}
		}
	}

	/**
	 * Renders editing the element.
	 *
	 * @param props The props.
	 *
	 * @returns {JSX.Element} The element to render.
	 */
	edit( props: RenderEditProps ): ReactElement | string {
		const label = this.options.label;
		const value = props.attributes[ this.options.name ] as string | string[];
		const options = arrayOrObjectToOptions( this.options.options );

		/**
		 * Function that is called whenever a new value is selected in the select element.
		 *
		 * @param event The change event.
		 */
		const onInput = ( event: React.ChangeEvent<HTMLSelectElement> ): void => {
			props.setAttributes( { [ this.options.name ]: event.target.value } );
		};

		return <div>
			<label htmlFor={ label }>{ label }</label>
			<select className="yoast-schema-select" id={ label } onInput={ onInput } value={ value }>
				{
					options.map(
						( option, index ) => <option key={ index } value={ option.value }>{ option.label }</option>,
					)
				}
			</select>
		</div>;
	}

	/**
	 * Adds the select to the block configuration.
	 *
	 * @returns The block configuration.
	 */
	configuration(): Partial<BlockConfiguration> {
		return {
			attributes: {
				[ this.options.name ]: {
					required: this.options.required === true,
				},
			},
		};
	}

	/**
	 * Checks if the instruction block is valid.
	 *
	 * @param blockInstance The attributes from the block.
	 *
	 * @returns `true` if the instruction block is valid, `false` if the block contains errors.
	 */
	valid( blockInstance: BlockInstance ): boolean {
		if ( this.options.required === true ) {
			return attributeExists( blockInstance, this.options.name as string ) &&
				attributeNotEmpty( blockInstance, this.options.name as string );
		}

		return attributeExists( blockInstance, this.options.name as string );
	}
}

BlockInstruction.register( "select", Select );
