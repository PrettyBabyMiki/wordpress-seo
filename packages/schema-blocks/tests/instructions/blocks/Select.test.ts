import * as renderer from "react-test-renderer";
import { shallow } from "enzyme";

import { ReactElement } from "@wordpress/element";
import { BlockConfiguration, BlockInstance } from "@wordpress/blocks";

import Select from "../../../src/instructions/blocks/Select";
import { RenderSaveProps } from "../../../src/core/blocks/BlockDefinition";
import { RenderEditProps } from "../../../src/core/blocks/BlockDefinition";

describe( "The Select instruction", () => {
	describe( "the save method", () => {
		it( "renders the correct React tree.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				options: {
					Korean: "korean",
					Tanzanian: "tanzanian",
					Australian: "australian",
				},
			};

			const props: RenderSaveProps = {
				attributes: {
					cuisine: "korean",
				},
			};

			const selectInstruction = new Select( 123, options );

			const tree = renderer
				.create( selectInstruction.save( props ) as ReactElement )
				.toJSON();

			expect( tree ).toMatchSnapshot();
		} );

		it( "renders the correct React tree when the selection options are an array.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				options: [
					"korean",
					"tanzanian",
					"australian",
				],
			};

			const props: RenderEditProps = {
				className: "",
				isSelected: false,
				setAttributes: jest.fn(),
				attributes: {
					cuisine: "korean",
				},
			};

			const selectInstruction = new Select( 123, options );

			const tree = renderer
				.create( selectInstruction.save( props ) as ReactElement )
				.toJSON();

			expect( tree ).toMatchSnapshot();
		} );
	} );

	describe( "the edit method", () => {
		it( "renders the correct React tree.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				options: {
					Korean: "korean",
					Tanzanian: "tanzanian",
					Australian: "australian",
				},
			};

			const props: RenderEditProps = {
				className: "",
				isSelected: false,
				setAttributes: jest.fn(),
				attributes: {
					cuisine: "korean",
				},
			};

			const selectInstruction = new Select( 123, options );

			const tree = renderer
				.create( selectInstruction.edit( props ) as ReactElement )
				.toJSON();

			expect( tree ).toMatchSnapshot();
		} );

		it( "renders the correct React tree when the selection options are an array.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				options: [
					"korean",
					"tanzanian",
					"australian",
				],
			};

			const props: RenderEditProps = {
				className: "",
				isSelected: false,
				setAttributes: jest.fn(),
				attributes: {
					cuisine: "korean",
				},
			};

			const selectInstruction = new Select( 123, options );

			const tree = renderer
				.create( selectInstruction.edit( props ) as ReactElement )
				.toJSON();

			expect( tree ).toMatchSnapshot();
		} );

		it( "correctly registers an onInput method.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				options: [
					"korean",
					"tanzanian",
					"australian",
				],
			};

			const props: RenderEditProps = {
				className: "",
				isSelected: false,
				setAttributes: jest.fn(),
				attributes: {
					cuisine: "korean",
				},
			};

			const selectInstruction = new Select( 123, options );

			const tree = shallow( selectInstruction.edit( props ) as ReactElement );

			const selectElement = tree.find( "select" ).first();

			const onInputEvent = {
				target: {
					value: "australian",
				},
			};

			selectElement.simulate( "input", onInputEvent );

			expect( props.setAttributes ).toBeCalledWith( { cuisine: "australian" } );
		} );
	} );

	describe( "the configuration method", () => {
		it( "returns the correct configuration.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				required: true,
				options: [
					"korean",
					"tanzanian",
					"australian",
				],
			};

			const expectedConfiguration: Partial<BlockConfiguration> = {
				attributes: {
					cuisine: {
						required: true,
					},
				},
			};

			const selectInstruction = new Select( 123, options );

			expect( selectInstruction.configuration() ).toEqual( expectedConfiguration );
		} );
	} );

	describe( "the valid method", () => {
		it( "returns true when the instruction is required and the value exists and is filled in.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				required: true,
				options: [
					"korean",
					"tanzanian",
					"australian",
				],
			};

			const selectInstruction = new Select( 123, options );

			const blockInstance: BlockInstance = {
				name: "select-instruction",
				clientId: "abcd-1234",
				isValid: true,
				innerBlocks: [],
				attributes: {
					cuisine: "tanzanian",
				},
			};

			expect( selectInstruction.valid( blockInstance ) ).toEqual( true );
		} );

		it( "returns true when the instruction is not required and the value exists.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				required: false,
				options: [
					"korean",
					"tanzanian",
					"australian",
				],
			};

			const selectInstruction = new Select( 123, options );

			const blockInstance: BlockInstance = {
				name: "select-instruction",
				clientId: "abcd-1234",
				isValid: true,
				innerBlocks: [],
				attributes: {
					cuisine: "",
				},
			};

			expect( selectInstruction.valid( blockInstance ) ).toEqual( true );
		} );

		it( "returns false when the instruction is not required and the value does not exist.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				required: false,
				options: [
					"korean",
					"tanzanian",
					"australian",
				],
			};

			const selectInstruction = new Select( 123, options );

			const blockInstance: BlockInstance = {
				name: "select-instruction",
				clientId: "abcd-1234",
				isValid: true,
				innerBlocks: [],
				attributes: {},
			};

			expect( selectInstruction.valid( blockInstance ) ).toEqual( false );
		} );

		it( "returns false when the instruction is required and the value does not exist.", () => {
			const options: Select["options"] = {
				name: "cuisine",
				label: "Cuisine",
				required: true,
				options: [
					"korean",
					"tanzanian",
					"australian",
				],
			};

			const selectInstruction = new Select( 123, options );

			const blockInstance: BlockInstance = {
				name: "select-instruction",
				clientId: "abcd-1234",
				isValid: true,
				innerBlocks: [],
				attributes: {},
			};

			expect( selectInstruction.valid( blockInstance ) ).toEqual( false );
		} );
	} );
} );
