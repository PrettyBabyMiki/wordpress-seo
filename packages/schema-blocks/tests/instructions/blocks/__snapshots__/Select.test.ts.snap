// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Select instruction the edit method renders the correct React tree when the selection options are an array. 1`] = `
<div>
  <label
    htmlFor="Cuisine"
  >
    Cuisine
  </label>
  <select
    className="yoast-schema-select"
    id="Cuisine"
    onInput={[Function]}
    value="korean"
  >
    <option
      value="korean"
    >
      korean
    </option>
    <option
      value="tanzanian"
    >
      tanzanian
    </option>
    <option
      value="australian"
    >
      australian
    </option>
  </select>
</div>
`;

exports[`The Select instruction the edit method renders the correct React tree. 1`] = `
<div>
  <label
    htmlFor="Cuisine"
  >
    Cuisine
  </label>
  <select
    className="yoast-schema-select"
    id="Cuisine"
    onInput={[Function]}
    value="korean"
  >
    <option
      value="korean"
    >
      Korean
    </option>
    <option
      value="tanzanian"
    >
      Tanzanian
    </option>
    <option
      value="australian"
    >
      Australian
    </option>
  </select>
</div>
`;

exports[`The Select instruction the save method renders the correct React tree when the selection options are an array. 1`] = `
<p
  data-id="cuisine"
  data-value="korean"
>
  <b>
    Cuisine
    :
  </b>
   
  korean
</p>
`;

exports[`The Select instruction the save method renders the correct React tree. 1`] = `
<p
  data-id="cuisine"
  data-value="korean"
>
  <b>
    Cuisine
    :
  </b>
   
  Korean
</p>
`;
